lib/pf/api/queue_cluster.pm

```diff
@@ -0,0 +1,285 @@
+    return;
+}
+
+sub cluster_notify {
+    my ($self, $method, @args) = @_;
+    foreach my $server ($self->servers) {
+        if ($server->{host} eq $pf::cluster::host_id) {
+            return $self->local_notify($method, @args);
+        }
+
+        if ($self->server_notify($server, $method, @args)) {
+            get_logger->debug("sent $method to $server->{host}");
+            last;
+        }
+    }
+    return; 
+}
+
+=head2 server_notify
+
+Sends a queue notify request to a server
+
+=cut
+
+sub server_notify {
+    my ($self, $server, $method, @args) = @_;
+    return $self->do_jsonrpc_notify($server, queue_submit => $self->queue, api => [$method, @args]);
+}
+
+=head2 mark_server_as_down
+
+Mark server as down
+
+=cut
+
+sub mark_server_as_down {
+    my ($self, $server) = @_;
+    $CHI_CACHE->set($server->{management_ip}, 1, {expires_in => $self->server_recheck_timeout});
+}
+
+=head2 is_server_down
+
+Checks if server is down
+
+=cut
+
+sub is_server_down {
+    my ($server) = @_;
+    return $CHI_CACHE->get($server->{management_ip});
+}
+
+sub notify_delayed {
+    my ($self, $delay, $method, @args) = @_;
+    if ($cluster_enabled) {
+        $self->cluster_notify_delayed($delay, $method, @args);
+    } else {
+        $self->local_notify_delayed($delay, $method, @args);
+    }
+    return;
+}
+
+=head2 cluster_notify_delayed
+
+Sends a queue notify delayed request to a server
+
+=cut
+
+sub cluster_notify_delayed {
+    my ($self, $delay, $method, @args) = @_;
+    foreach my $server ($self->servers) {
+        if ($server->{host} eq $pf::cluster::host_id) {
+            return $self->local_notify_delayed($delay, $method, @args);
+        }
+
+        if ($self->server_notify_delayed($server, $delay, $method, @args)) {
+            get_logger->info("sent $method to $server->{host}");
+            last;
+        }
+    }
+    return; 
+}
+
+=head2 server_notify_delayed
+
+server_notify_delayed
+
+=cut
+
+sub server_notify_delayed {
+    my ($self, $server, $delay, $method, @args) = @_;
+    return $self->do_jsonrpc_notify($server, queue_submit_delayed => $self->queue, $delay, api => [$method, @args]);
+}
+
+=head2 do_jsonrpc_notify
+
+do_jsonrpc_notify
+
+=cut
+
+sub do_jsonrpc_notify {
+    my ($self, $server, $method, @args) = @_;
+    my $apiclient = $self->jsonrpcclient($server);
+    my $results = $apiclient->notify($method, @args);
+    unless ($results) {
+        my $dead_for = $self->server_recheck_timeout;
+        get_logger->error("Failed to send $method to $server->{host} marking as dead for $dead_for seconds");
+        $self->mark_server_as_down($server);
+    }
+
+    return $results;
+}
+
+=head2 jsonrpcclient
+
+jsonrpcclient
+
+=cut
+
+sub jsonrpcclient {
+    my ($self, $server) = @_;
+    require pf::api::jsonrpcclient;
+    return pf::api::jsonrpcclient->new(
+        host => $server->{management_ip},
+        proto => 'https',
+        %{$self->jsonrpc_args // {}},
```

Comment:
Already defaulting to '{}'

Reply:
That is there to if jsonrpc_args was set to undef.